connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetrics.java

```diff
@@ -0,0 +1,227 @@
+     *
+     * @param groupName    the name of the metric group; may not be null
+     * @param tagKeyValues pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public MetricGroup group(String groupName, String... tagKeyValues) {
+        return group(groupName, false, tagKeyValues);
+    }
+
+    /**
+     * Get or create a {@link MetricGroup} with the specified group name and the given tags.
+     *
+     * @param groupName       the name of the metric group; may not be null
+     * @param includeWorkerId true if the tags should include the worker ID
+     * @param tagKeyValues    pairs of tag name and values
+     * @return the {@link MetricGroup} that can be used to create metrics; never null
+     */
+    public synchronized MetricGroup group(String groupName, boolean includeWorkerId, String... tagKeyValues) {
+        MetricGroup group = groupsByName.get(groupName);
+        if (group == null) {
+            Map<String, String> tags = tags(includeWorkerId ? workerId : null, tagKeyValues);
+            group = new MetricGroup(groupName, tags);
+            MetricGroup previous = groupsByName.putIfAbsent(groupName, group);
+            if (previous != null) group = previous;
+        }
+        return group;
+    }
+
+    /**
+     * Get the time.
+     *
+     * @return the time; never null
+     */
+    public Time time() {
+        return time;
+    }
+
+    /**
+     * Stop and unregister the metrics from any reporters.
+     */
+    public void stop() {
+        metrics.close();
+        LOG.debug("Unregistering Connect metrics with JMX for worker '{}'", workerId);
+        AppInfoParser.unregisterAppInfo(JMX_PREFIX, workerId);
+    }
+
+    /**
+     * A group of metrics. Each group maps to a JMX MBean and each metric maps to an MBean attribute.
+     */
+    public class MetricGroup {
+        private final String groupName;
+        private final Map<String, String> tags;
+
+        protected MetricGroup(String groupName, Map<String, String> tags) {
+            this.groupName = groupName;
+            this.tags = tags;
+        }
+
+        /**
+         * Create the name of a metric that belongs to this group and has the group's tags.
+         *
+         * @param name the name of the metric/attribute; may not be null
+         * @param desc the description for the metric/attribute; may not be null
+         * @return the metric name; never null
+         */
+        public MetricName metricName(String name, String desc) {
+            return metrics.metricName(name, groupName, desc, tags);
+        }
+
+        /**
+         * The {@link Metrics} that this group belongs to.
+         *
+         * @return the metrics; never null
+         */
+        public Metrics metrics() {
+            return metrics;
+        }
+
+        Map<String, String> tags() {
+            return Collections.unmodifiableMap(tags);
```

Comment:
you can just do the conversion to unmodifiable map one time in the constructor. it looks like at the moment this is only accessed in tests anyway.

Reply:
ack